specs/core/0_beacon-chain.md

```diff
@@ -1513,54 +1266,319 @@ def penalize_validator(state: BeaconState, index: ValidatorIndex) -> None:
+* Let `justified_head` be the descendant of `finalized_head` with the highest epoch that has been justified for at least 1 epoch. (A block `B` is justified if there is a descendant of `B` in `store` the processing of which sets `B` as justified.) If no such descendant exists set `justified_head` to `finalized_head`.
+* Let `get_ancestor(store: Store, block: BeaconBlock, slot: SlotNumber) -> BeaconBlock` be the ancestor of `block` with slot number `slot`. The `get_ancestor` function can be defined recursively as `def get_ancestor(store: Store, block: BeaconBlock, slot: SlotNumber) -> BeaconBlock: return block if block.slot == slot else get_ancestor(store, store.get_parent(block), slot)`.
+* Let `get_latest_attestation(store: Store, validator: Validator) -> Attestation` be the attestation with the highest slot number in `store` from `validator`. If several such attestations exist, use the one the [validator](#dfn-validator) `v` observed first.
+* Let `get_latest_attestation_target(store: Store, validator: Validator) -> BeaconBlock` be the target block in the attestation `get_latest_attestation(store, validator)`.
+* Let `get_children(store: Store, block: BeaconBlock) -> List[BeaconBlock]` returns the child blocks of the given `block`.
+* Let `justified_head_state` be the resulting `BeaconState` object from processing the chain up to the `justified_head`.
+* The `head` is `lmd_ghost(store, justified_head_state, justified_head)` where the function `lmd_ghost` is defined below. Note that the implementation below is suboptimal; there are implementations that compute the head in time logarithmic in slot count.
+
+```python
+def lmd_ghost(store: Store, start_state: BeaconState, start_block: BeaconBlock) -> BeaconBlock:
+    validators = start_state.validator_registry
+    active_validators = [
+        validators[i]
+        for i in get_active_validator_indices(validators, start_state.slot)
+    ]
+    attestation_targets = [
+        get_latest_attestation_target(store, validator)
+        for validator in active_validators
+    ]
+
+    def get_vote_count(block: BeaconBlock) -> int:
+        return len([
+            target
+            for target in attestation_targets
+            if get_ancestor(store, target, block.slot) == block
+        ])
+
+    head = start_block
+    while 1:
+        children = get_children(store, head)
+        if len(children) == 0:
+            return head
+        head = max(children, key=get_vote_count)
+```
+
+## Beacon chain state transition function
+
+We now define the state transition function. This is a parent to child block transition. At a high level the state transition is made up of three parts:
+
+1. The per-slot transitions, which happens every slot.
+2. The per-block transitions, which happens at every block.
+3. The per-epoch transitions, which happens at every epoch boundary (i.e. `(state.slot + 1) % EPOCH_LENGTH == 0`).
```

Comment:
"which happens at every epoch boundary" => More specifically, at the last slot of every epoch.